src/Microsoft.Health.Dicom.Core/Extensions/ConfigurationBinderExtensions.cs

```diff
@@ -0,0 +1,138 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+using EnsureThat;
+using Microsoft.Extensions.Configuration;
+
+namespace Microsoft.Health.Dicom.Core.Extensions;
+
+internal static class ConfigurationBinderExtensions
+{
+    public static void Set<T>(this IConfiguration configuration, T value) where T : class
+        => Set(configuration, typeof(T), value, null); // TODO: Avoid boxing?
+
+    public static void Set<T>(this IConfiguration configuration, T value, Action<BinderOptions> configureOptions) where T : class
+        => Set(configuration, typeof(T), value, configureOptions);
+
+    public static void Set(this IConfiguration configuration, Type type, object value)
+        => Set(configuration, type, value, null);
+
+    public static void Set(this IConfiguration configuration, Type type, object value, Action<BinderOptions> configureOptions)
+    {
+        EnsureArg.IsNotNull(configuration, nameof(configuration));
+        EnsureArg.IsNotNull(type, nameof(type));
+        EnsureArg.IsNotNull(value, nameof(value));
+
+        var options = new BinderOptions();
+        configureOptions?.Invoke(options);
+
+        // TODO: Cache a dynamic method to perform the reflection only once. The Get methods are not cached,
+        //       so it is not unreasonable to leave this as-is
+
+        // TODO: Support copying from IConfiguration/IConfigurationSection
+
+        // TODO: Support value types and primitives
+        if (type.IsValueType)
+            throw new InvalidOperationException();
+
+        CopyToConfiguration(configuration, type, value, options);
+    }
+
+    private static void CopyToConfiguration(IConfiguration configuration, Type type, object value, BinderOptions options)
+    {
+        if (value == null)
+            return;
+
+        if (IsLiteralType(type))
+        {
+            if (configuration is not IConfigurationSection section)
+                throw new InvalidOperationException(
+                        string.Format(CultureInfo.CurrentCulture, DicomCoreResource.InvalidTypeBinding, type.Name));
+
+            // Special-case DateTime types to use round-trip
+            if (type == typeof(DateTime) || type == typeof(DateTimeOffset))
+            {
+                section.Value = ((IFormattable)value).ToString("O", CultureInfo.InvariantCulture);
+            }
+            else
+            {
+                TypeConverter converter = TypeDescriptor.GetConverter(type);
+                section.Value = converter.ConvertToInvariantString(value);
+            }
+        }
+        else if (type.IsArray || IsArrayLike(type))
+        {
+            Type elementType;
+            if (type.IsArray)
+            {
+                elementType = type.GetElementType();
+                if (type.IsArray && type.GetArrayRank() > 1)
```

Comment:
Do we need to check `type.IsArray` again here?

Reply:
Whoops! Nope